iotdb-core/consensus/src/main/java/org/apache/iotdb/consensus/ratis/DiskGuardian.java

```diff
@@ -0,0 +1,246 @@
+        workerThread, this::snapshotDaemon, 0L, daemonIntervalMs, TimeUnit.SECONDS);
+
+    // then schedule all checker daemons
+    snapshotArbitrators.forEach(
+        (interval, checkers) ->
+            ScheduledExecutorUtil.safelyScheduleWithFixedDelay(
+                workerThread,
+                () -> checkerDaemon(checkers),
+                0L,
+                interval.toLong(TimeUnit.SECONDS),
+                TimeUnit.SECONDS));
+  }
+
+  void stop() throws InterruptedException {
+    if (isStopped.compareAndSet(true, false)) {
+      workerThread.shutdown();
+      workerThread.awaitTermination(5, TimeUnit.SECONDS);
+    }
+  }
+
+  /** call this method to register checkers before {@link #start()} */
+  void registerChecker(Predicate<RaftLogSummary> checker, TimeDuration interval) {
+    final List<Predicate<RaftLogSummary>> checkers =
+        snapshotArbitrators.computeIfAbsent(interval, i -> new CopyOnWriteArrayList<>());
+    checkers.add(checker);
+  }
+
+  /** periodically woken up. Will take a snapshot if a flag is set. */
+  private void snapshotDaemon() {
+    if (isStopped.get()) {
+      return;
+    }
+    for (RaftGroupId groupId : serverRef.get().getServer().getGroupIds()) {
+      if (getSnapshotFlag(groupId).get()) {
+        try {
+          serverRef.get().triggerSnapshot(Utils.fromRaftGroupIdToConsensusGroupId(groupId));
+          final boolean flagCleared = snapshotFlag.get(groupId).compareAndSet(true, false);
+          Preconditions.assertTrue(flagCleared);
+        } catch (ConsensusException e) {
+          logger.warn(
+              "{} take snapshot failed for group {} due to {}. Disk file status {}",
+              this,
+              groupId,
+              e,
+              getLatestSummary(groupId).orElse(null));
+        }
+      }
+    }
+  }
+
+  /** periodically woken up. Will check the snapshot condition and set the snapshot flag. */
+  private void checkerDaemon(List<Predicate<RaftLogSummary>> checkerList) {
+    if (isStopped.get()) {
+      return;
+    }
+    for (RaftGroupId groupId : serverRef.get().getServer().getGroupIds()) {
+      final Optional<RaftLogSummary> summary = getLatestSummary(groupId);
+      if (summary.isPresent()) {
+        final Optional<Boolean> anyCheckerPositive =
+            checkerList.stream()
+                .map(checker -> checker.test(summary.get()))
+                .filter(Boolean::booleanValue)
+                .findAny();
+        if (anyCheckerPositive.isPresent()) {
+          getSnapshotFlag(groupId).set(true);
+        }
+      }
+    }
+  }
+
+  private AtomicBoolean getSnapshotFlag(RaftGroupId groupId) {
+    return snapshotFlag.computeIfAbsent(groupId, id -> new AtomicBoolean(false));
+  }
+
+  private Optional<RaftLogSummary> getLatestSummary(RaftGroupId groupId) {
+    // initialize the RaftLog Summary for the first time
+    if (!bookkeeper.containsKey(groupId)) {
+      synchronized (bookkeeper) {
```

Comment:
use compute API to ensure atomic? You can ask ChatGPT to optimize your code here

Reply:
Good to know.